### PR TITLE
[LogitsProcs] Deduplicate built-in LP implementation logic

### PR DESCRIPTION
**Pull request by njhill**
The state-updating for sparse LPs is common and can be shared.

### CODE REVIEW

File: examples/offline_inference/logits_processor.py [hunk 1]
```diff
@@ -60,31 +60,14 @@ def is_argmax_invariant(self) -> bool:
...
-                b_val = self.req_info.pop(bdx, None)
-                if a_val is not None:
-                    self.req_info[bdx] = a_val
-                if direct == MoveDirectionality.SWAP and b_val is not None:
-                    self.req_info[adx] = b_val
+        process_dict_updates(
+            self.req_info,
+            batch_update,
+            # This function returns the LP's state based on the request details,
+            # or None if this LP does not apply to the request.
```
Comment (afeldman-nm) [inline examples/offline_inference/logits_processor.py]: Which "function" is referred to by these lines? `process_dict_updates()`, or the `lambda params`? If it is the former - doesn't this comment belong inside the definition of `process_dict_updates()`? If it is the latter - it would seem that this comment does not describe the lambda?

File: vllm/v1/sample/logits_processor/builtin.py [hunk 2]
```diff
@@ -208,52 +177,18 @@ def is_argmax_invariant(self) -> bool:
...
+    ) -> Optional[tuple[int, Sequence[int], set[int]]]:
+        min_tokens = params.min_tokens
+        if not min_tokens or len(output_tok_ids) >= min_tokens:
+            return None
+        return min_tokens, output_tok_ids, params.all_stop_token_ids
 
+    def update_state(self, batch_update: Optional[BatchUpdate]):
+        needs_update = process_dict_updates(self.min_toks, batch_update,
+                                            self.add_request)
         if self.min_toks:
```
Comment (afeldman-nm) [inline vllm/v1/sample/logits_processor/builtin.py:200]: It is great to have this helper function `process_dict_updates()` which deduplicates the logic of managing state using a dict. However, I am wondering if this is the largest possible granularity at which we can deduplicate. For example I wonder if it would be possible to have a `SparseStateLogitsProcessor(LogitsProcessor)` subclass which handles basically every aspect of operating a logits processor with sparse (dict-based) state? This subclass could in turn be subclassed to implement specific logits processors with dict-based state. The `SparseStateLogitsProcessor` `update_state()` method could look something like

``` python
    def update_state(self, batch_update: Optional[BatchUpdate]):
        """State-update logic common to all `SparseStateLogitsProcessors`
        If we take this approach, the contents of `process_dict_updates()` could probably be moved into this method
        
        Note that I have replaced `self.min_toks` with `self.state_dict` to reflect that all `SparseStateLogitsProcessor`'s have a state dict. I am probably omitting some details here about how `self.state_dict` could be handled in a generic way which allows subclasses to control the state dict value type
        """
        needs_update = process_dict_updates(self.state_dict, batch_update,
                                            self.add_request)
        
        # Invoke subclass-specific state handling
        self._update_state(batch_update, needs_update)
        
    def _update_state(batch_update, needs_update):
        """Subclass-specific state updates"""
    
        if self.state_dict:
            # Check for any requests that have attained their min tokens.
            to_remove = tuple(index for index, (min_toks, out_tok_ids,
                                                _) in self.state_dict.items()
                            if len(out_tok_ids) >= min_toks)
            if to_remove:
                needs_update = True
                for index in to_remove:
                    del self.state_dict[index]

        # Update tensors if needed.
        if needs_update:
            reqs: list[int] = []
            tok_ids: list[int] = []
            for req, (_, _, stop_tok_ids) in self.state_dict.items():
                reqs.extend([req] * len(stop_tok_ids))
                tok_ids.extend(stop_tok_ids)
            self.logits_slice = (self._device_tensor(reqs, torch.int32),
                                self._device_tensor(tok_ids, torch.int32))
```

Do you think this could work?

**Review by afeldman-nm — CHANGES_REQUESTED** [review]
Thanks for this Nick. Left a few comments.

File: examples/offline_inference/logits_processor.py (hunk 1 above)
Comment (njhill) [reply examples/offline_inference/logits_processor.py]: It's the latter. It does describe it.. the state in this case is the target token (an int). If there aren't extra_args or tehy don't contain it then the lambda will return `None`. I should probably change the wording to "the LP's per-request state".

I just noticed though, the delclared type of `self.req_info` in the `__init__` method appears to be wrong (`dict[int, SamplingParams]` should be `dict[int, int]` I think?

File: vllm/v1/sample/logits_processor/builtin.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/sample/logits_processor/builtin.py:200]: Yes I had the same thought, but would like to get this simplification as a first step towards that.

I was actually thinking it would be good to make this the basis of all LP's, so they don't have to deal with the batch reordering logic. And then even the dense ones would just recreate the arrays/tensors from the dict when needed.

However, forthcoming changes to how the persistent batch works may also affect things here, so I'd like to figure out the interface changes taking that into account too.

Comment (njhill) [reply vllm/v1/sample/logits_processor/builtin.py:200]: W.r.t. your example, I was thinking pretty much exactly the same, except the simpler _update_states method wouldn't need to take 2 args.. it would just be passed `None` rather than the original batch_update if `needs_update` is `False`.

Again would like to get this PR merged as-is first though.

File: examples/offline_inference/logits_processor.py (hunk 1 above)
Comment (afeldman-nm) [reply examples/offline_inference/logits_processor.py]: Okay makes sense. And yes woops I thought I fixed all of those type annotations.

**Review by afeldman-nm — APPROVED** [review]

**Review by DarkLight1337 — APPROVED** [review]
